src/ucp/proto/proto_debug.c

```diff
@@ -382,6 +430,207 @@ void ucp_proto_select_info_str(ucp_worker_h worker,
+    return perf_node;
+}
+
+ucp_proto_perf_node_t *ucp_proto_perf_node_new_select(const char *name,
+                                                      unsigned selected_child,
+                                                      const char *desc_fmt, ...)
+{
+    ucp_proto_perf_node_t *perf_node;
+
+    perf_node = UCP_PROTO_PERF_NODE_NEW(SELECT, name, desc_fmt);
+    perf_node->selected_child = selected_child;
+    return perf_node;
+}
+
+ucp_proto_perf_node_t *
+ucp_proto_perf_node_new_compose(const char *name, const char *desc_fmt, ...)
+{
+    return UCP_PROTO_PERF_NODE_NEW(COMPOSE, name, desc_fmt);
+}
+
+void ucp_proto_perf_node_ref(ucp_proto_perf_node_t *perf_node)
+{
+    if (perf_node != NULL) {
+        ++perf_node->refcount;
+    }
+}
+
+void ucp_proto_perf_node_deref(ucp_proto_perf_node_t **perf_node_p)
+{
+    ucp_proto_perf_node_t *perf_node = *perf_node_p;
+
+    if (perf_node == NULL) {
+        return;
+    }
+
+    ucs_assertv(perf_node->refcount > 0, "perf_node=%p name='%s' desc='%s'",
+                perf_node, perf_node->name, perf_node->desc);
+    --perf_node->refcount;
+    if (perf_node->refcount == 0) {
+        ucp_proto_perf_node_free(perf_node);
+        *perf_node_p = NULL;
+    }
+}
+
+static void
+ucp_proto_perf_node_append_child(ucp_proto_perf_node_t *perf_node,
+                                 ucp_proto_perf_node_t *child_perf_node)
+{
+    ucs_status_t status;
+
+    status = ucs_array_append(ucp_proto_perf_node, &perf_node->children);
+    if (status != UCS_OK) {
+        ucs_diag("failed to add perf node child");
+        return;
+    }
+
+    *ucs_array_last(&perf_node->children) = child_perf_node;
+}
+
+void ucp_proto_perf_node_own_child(ucp_proto_perf_node_t *perf_node,
+                                   ucp_proto_perf_node_t **child_perf_node_p)
+{
+    if (*child_perf_node_p == NULL) {
+        return;
+    } else if (perf_node == NULL) {
+        ucp_proto_perf_node_deref(child_perf_node_p);
+        return;
+    }
+
+    ucp_proto_perf_node_append_child(perf_node, *child_perf_node_p);
+}
+
+void ucp_proto_perf_node_add_child(ucp_proto_perf_node_t *perf_node,
+                                   ucp_proto_perf_node_t *child_perf_node)
+{
+    if ((perf_node == NULL) || (child_perf_node == NULL)) {
+        return;
+    }
+
+    ucp_proto_perf_node_append_child(perf_node, child_perf_node);
+    ucp_proto_perf_node_ref(child_perf_node);
+}
+
+void ucp_proto_perf_node_add_data(ucp_proto_perf_node_t *perf_node,
+                                  const char *name,
+                                  const ucs_linear_func_t value)
+{
+    ucp_proto_perf_node_data_t *data;
+    ucs_status_t status;
+
+    if (perf_node == NULL) {
+        return;
+    }
+
+    ucs_assert(perf_node->type == UCP_PROTO_PERF_NODE_TYPE_DATA);
+
+    status = ucs_array_append(ucp_proto_perf_node_data, &perf_node->data);
+    if (status != UCS_OK) {
+        ucs_diag("failed to add perf node data");
+        return;
+    }
```

Comment:
maybe replace by `ucs_assertv_always(status != UCS_OK, "failed to add perf node data: %s", ucs_status_string(status))`
normally it shouldn't happen

Reply:
same here, in perf_node we ignore errors related to memory allocation since the applications can work without having a performance information tree.